File: trunk/src/main/java/net/esper/filter/FilterSpecParamIn.java
package net.esper.filter;

import net.esper.event.EventType;
import net.esper.pattern.MatchedEventMap;
import net.esper.collection.MultiKeyUntyped;
import net.esper.util.JavaClassHelper;

import java.util.Map;
import java.util.Arrays;
import java.util.List;

/**
 * This class represents a 'in' filter parameter in an {@link net.esper.filter.FilterSpec} filter specification.
 * <p>
 * The 'in' checks for a list of values.
 */
public final class FilterSpecParamIn extends FilterSpecParam
{
    private final List<FilterSpecParamInValue> listOfValues;
    private MultiKeyUntyped inListConstantsOnly;
    private Class propertyType;

    /**
     * Ctor.
     * @param propertyName is the event property name
     * @param filterOperator is expected to be the IN-list operator
     * @param listofValues is a list of constants and event property names
     * @param isAllConstants true if there are only constants, nd false if there is one or more property in the values
     * @param propertyType is the type of the property
     * @throws IllegalArgumentException for illegal args
     */
    public FilterSpecParamIn(String propertyName,
                             FilterOperator filterOperator,
                             List<FilterSpecParamInValue> listofValues,
                             boolean isAllConstants,
                             Class propertyType)
        throws IllegalArgumentException
    {
        super(propertyName, filterOperator);

        this.propertyType = propertyType;
        this.listOfValues = listofValues;
        if (isAllConstants)
        {
            Object[] constants = new Object[listOfValues.size()];
            int count = 0;
            for (FilterSpecParamInValue valuePlaceholder : listOfValues)
            {
                constants[count++] = valuePlaceholder.getFilterValue(null);
            }
            inListConstantsOnly = new MultiKeyUntyped(constants);
        }

        if (!(filterOperator == FilterOperator.IN_LIST_OF_VALUES))
        {
            throw new IllegalArgumentException("Illegal filter operator " + filterOperator + " supplied to " +
                    "in-values filter parameter");
        }
    }

    public final Class getFilterValueClass(Map<String, EventType> taggedEventTypes)
    {
        for (FilterSpecParamInValue valuePlaceholder : listOfValues)
        {
            Class clazz = valuePlaceholder.validate(taggedEventTypes);
            Class clazzBoxed = JavaClassHelper.getBoxedType(clazz);
            Class propBoxed = JavaClassHelper.getBoxedType(propertyType); 
            if (clazzBoxed != propBoxed)
            {
                throw new IllegalStateException("Automatic conversion of type " + clazz + " to type " + propertyType + " not allowed");
            }
        }

        return MultiKeyUntyped.class;
    }

    public final Object getFilterValue(MatchedEventMap matchedEvents)
    {
        // If the list of values consists of all-constants and no event properties, then use cached version
        if (inListConstantsOnly != null)
        {
            return inListConstantsOnly;
        }

        // Determine actual values since the in-list of values contains one or more event properties
        Object[] actualValues = new Object[listOfValues.size()];
        int count = 0;
        for (FilterSpecParamInValue valuePlaceholder : listOfValues)
        {
            actualValues[count++] = valuePlaceholder.getFilterValue(matchedEvents);
        }
        return new MultiKeyUntyped(actualValues);
    }

    public List<FilterSpecParamInValue> getListOfValues()
    {
        return listOfValues;
    }

    public final String toString()
    {
        return super.toString() + "  in=(listOfValues=" + listOfValues.toString() + ")";
    }

    public boolean equals(Object obj)
    {
        if (this == obj)
        {
            return true;
        }

        if (!(obj instanceof FilterSpecParamIn))
        {
            return false;
        }

        FilterSpecParamIn other = (FilterSpecParamIn) obj;
        if (!super.equals(other))
        {
            return false;
        }

        if (listOfValues.size() != other.listOfValues.size())
        {
            return false;
        }

        if (!(Arrays.deepEquals(listOfValues.toArray(), other.listOfValues.toArray())))
        {
            return false;
        }
        return true;
    }
}

File: NEsper/filter/FilterSpecParamIn.cs
///////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
///////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;

using net.esper.compat;
using net.esper.collection;
using net.esper.pattern;

namespace net.esper.filter
{
    /// <summary>
	/// This class represents a 'in' filter parameter in an {@link net.esper.filter.FilterSpecCompiled} filter specification.
    /// <para>
	/// The 'in' checks for a list of values.
	/// </para>
	/// </summary>
	public sealed class FilterSpecParamIn : FilterSpecParam
	{
	    private readonly IList<FilterSpecParamInValue> listOfValues;
	    private MultiKeyUntyped inListConstantsOnly;

	    /// <summary>Ctor.</summary>
	    /// <param name="propertyName">is the event property name</param>
	    /// <param name="filterOperator">is expected to be the IN-list operator</param>
	    /// <param name="listofValues">is a list of constants and event property names</param>
	    /// <throws>ArgumentException for illegal args</throws>
	    public FilterSpecParamIn(String propertyName,
	                             FilterOperator filterOperator,
	                             IList<FilterSpecParamInValue> listofValues)
	    	: base(propertyName, filterOperator)
	    {
	        this.listOfValues = listofValues;

	        bool isAllConstants = false;
	        foreach (FilterSpecParamInValue value in listofValues)
	        {
	            if (value is InSetOfValuesEventProp)
	            {
	                isAllConstants = false;
	                break;
	            }
	        }

	        if (isAllConstants)
	        {
	            Object[] constants = new Object[listOfValues.Count];
	            int count = 0;
	            foreach (FilterSpecParamInValue valuePlaceholder in listOfValues)
	            {
	                constants[count++] = valuePlaceholder.GetFilterValue(null);
	            }
	            inListConstantsOnly = new MultiKeyUntyped(constants);
	        }

	        if ((filterOperator != FilterOperator.IN_LIST_OF_VALUES) && ((filterOperator != FilterOperator.NOT_IN_LIST_OF_VALUES)))
	        {
	            throw new ArgumentException("Illegal filter operator " + filterOperator + " supplied to " +
	                    "in-values filter parameter");
	        }
	    }

        /// <summary>
        /// Return the filter parameter constant to filter for.
        /// </summary>
        /// <param name="matchedEvents">is the prior results that can be used to determine filter parameters</param>
        /// <returns>filter parameter constant's value</returns>
	    public override Object GetFilterValue(MatchedEventMap matchedEvents)
	    {
	        // If the list of values consists of all-constants and no event properties, then use cached version
	        if (inListConstantsOnly != null)
	        {
	            return inListConstantsOnly;
	        }

	        // Determine actual values since the in-list of values contains one or more event properties
	        Object[] actualValues = new Object[listOfValues.Count];
	        int count = 0;
	        foreach (FilterSpecParamInValue valuePlaceholder in listOfValues)
	        {
	            actualValues[count++] = valuePlaceholder.GetFilterValue(matchedEvents);
	        }
	        return new MultiKeyUntyped(actualValues);
	    }

        /// <summary>
        /// Returns the list of values we are asking to match.
        /// </summary>
        /// <returns>list of filter values</returns>
	    public IList<FilterSpecParamInValue> ListOfValues
	    {
            get { return listOfValues; }
	    }

        /// <summary>
        /// Returns a <see cref="T:System.String"></see> that represents the current <see cref="T:System.Object"></see>.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.String"></see> that represents the current <see cref="T:System.Object"></see>.
        /// </returns>
	    public override String ToString()
	    {
	        return base.ToString() + "  in=(listOfValues=" + listOfValues.ToString() + ')';
	    }

        /// <summary>
        /// Determines whether the specified <see cref="T:System.Object"></see> is equal to the current <see cref="T:System.Object"></see>.
        /// </summary>
        /// <param name="obj">The <see cref="T:System.Object"></see> to compare with the current <see cref="T:System.Object"></see>.</param>
        /// <returns>
        /// true if the specified <see cref="T:System.Object"></see> is equal to the current <see cref="T:System.Object"></see>; otherwise, false.
        /// </returns>
	    public override bool Equals(Object obj)
	    {
	        if (this == obj)
	        {
	            return true;
	        }

	        if (!(obj is FilterSpecParamIn))
	        {
	            return false;
	        }

	        FilterSpecParamIn other = (FilterSpecParamIn) obj;
	        if (!base.Equals(other))
	        {
	            return false;
	        }

	        if (listOfValues.Count != other.listOfValues.Count)
	        {
	            return false;
	        }

	        if (!CollectionHelper.AreEqual(listOfValues, other.listOfValues))
	        {
	            return false;
	        }
	        return true;
	    }

        /// <summary>
        /// Serves as a hash function for a particular type.
        /// </summary>
        /// <returns>
        /// A hash code for the current <see cref="T:System.Object"></see>.
        /// </returns>
		public override int GetHashCode()
		{
			return base.GetHashCode();
		}
	}
} // End of namespace
